In [1]:
import pandas as pd

df = pd.read_csv(
    '/Users/benedikt/Documents/code/data-science-g5/woche1/airbnb/data/raw/Airbnb_Open_Data.csv')

/var/folders/t5/m580657s5rxblthmnv5y6vfm0000gn/T/ipykernel_53032/1238865731.py:3: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(


**Missing Values/ Invalid Values**
- Potenziell ungültige Werte 
    - minimum nights - negative Werte
    - availability 365 - negative Werte
- Spalte Licence ist praktisch leer (und nur 2 unique values)

**Spaltentyp & Formatierung**
- Preise sind als String gespeichert
- Preise haben ein $ vor der Zahl
- Preise haben "," als Tausender-Trennzeichen
- Viele Spalten haben den Typ "Object"


**Redundante oder nicht sinnvolle Spalten**
- Unbedeutende Spalten dabei

**Duplicates**
- Doppelte Werte


## Missing Values

In [2]:
df.describe()

id       host id            lat           long  \
count  1.025990e+05  1.025990e+05  102591.000000  102591.000000   
mean   2.914623e+07  4.925411e+10      40.728094     -73.949644   
std    1.625751e+07  2.853900e+10       0.055857       0.049521   
min    1.001254e+06  1.236005e+08      40.499790     -74.249840   
25%    1.508581e+07  2.458333e+10      40.688740     -73.982580   
50%    2.913660e+07  4.911774e+10      40.722290     -73.954440   
75%    4.320120e+07  7.399650e+10      40.762760     -73.932350   
max    5.736742e+07  9.876313e+10      40.916970     -73.705220   

       Construction year  minimum nights  number of reviews  \
count      102385.000000   102190.000000      102416.000000   
mean         2012.487464        8.135845          27.483743   
std             5.765556       30.553781          49.508954   
min          2003.000000    -1223.000000           0.000000   
25%          2007.000000        2.000000           1.000000   
50%          2012.000000        3.000000           7.000000   
75%          2017.000000        5.000000          30.000000   
max          2022.000000     5645.000000        1024.000000   

       reviews per month  review rate number  calculated host listings count  \
count       86720.000000       102273.000000                   102280.000000   
mean            1.374022            3.279106                        7.936605   
std             1.746621            1.284657                       32.218780   
min             0.010000            1.000000                        1.000000   
25%             0.220000            2.000000                        1.000000   
50%             0.740000            3.000000                        1.000000   
75%             2.000000            4.000000                        2.000000   
max            90.000000            5.000000                      332.000000   

       availability 365  
count     102151.000000  
mean         141.133254  
std          135.435024  
min          -10.000000  
25%            3.000000  
50%           96.000000  
75%          269.000000  
max         3677.000000

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  object 
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host name                       102193 non-null  object 
 5   neighbourhood group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

In [4]:
df.isna().sum()

id                                     0
NAME                                 250
host id                                0
host_identity_verified               289
host name                            406
neighbourhood group                   29
neighbourhood                         16
lat                                    8
long                                   8
country                              532
country code                         131
instant_bookable                     105
cancellation_policy                   76
room type                              0
Construction year                    214
price                                247
service fee                          273
minimum nights                       409
number of reviews                    183
last review                        15893
reviews per month                  15879
review rate number                   326
calculated host listings count       319
availability 365                     448
house_rules     

In [2]:
(df.isna().sum() / df.shape[0]) * 100

id                                 0.000000
NAME                               0.243667
host id                            0.000000
host_identity_verified             0.281679
host name                          0.395715
neighbourhood group                0.028265
neighbourhood                      0.015595
lat                                0.007797
long                               0.007797
country                            0.518524
country code                       0.127682
instant_bookable                   0.102340
cancellation_policy                0.074075
room type                          0.000000
Construction year                  0.208579
price                              0.240743
service fee                        0.266084
minimum nights                     0.398639
number of reviews                  0.178364
last review                       15.490404
reviews per month                 15.476759
review rate number                 0.317742
calculated host listings count  

Ideen:
-> Komplette Duplikate löschen

id                                 0.000000 -> Nichts zu tun
NAME                               0.243667 -> Wenn nicht gegeben, Zeile entfernen
host id                            0.000000 -> Nichts zu tun
host_identity_verified             0.281679 -> Zeilen entfernen oder durch "Unbekannt ersetzen"/ NA lassen
host name                          0.395715 -> Ignorieren oder durch "Unbekannt"/ NA ersetzen
neighbourhood group                0.028265 -> Ignorieren
neighbourhood                      0.015595 -> Ignorieren
lat                                0.007797 -> Auf Duplikate prüfen oder entfernen (Zeilen)
long                               0.007797 -> Auf Duplikate prüfen oder entfernen (Zeilen)
country                            0.518524 -> Irrelevant
country code                       0.127682 -> Irrelevant
instant_bookable                   0.102340 -> 
cancellation_policy                0.074075 -> Ignorieren
room type                          0.000000 -> Nichts zu tun
Construction year                  0.208579 -> Als Int behandeln, Imputation, Jahr aufgrund von lat/ long schätzen
price                              0.240743 -> Schätzen oder durch Modell bestimmen (Lineare Regression)
service fee                        0.266084 -> Wie vorheriges
minimum nights                     0.398639 -> ihr seid drann
number of reviews                  0.178364
last review                       15.490404
reviews per month                 15.476759
review rate number                 0.317742
calculated host listings count     0.310919
availability 365                   0.436651
house_rules                       50.810437
license                           99.998051

In [4]:
df.shape

(102599, 26)

In [7]:
df.drop_duplicates(subset=["lat", "long", "NAME", "id"]).shape

(102058, 26)

**Was können wir mit fehlenden Werten Machen?**
- Löschen
  - Zeile oder Spalte
  - Kann zu Verzerrungen führen -> Prüfen, ob die Zeilen mit NA Spezialfälle sind

In [6]:
df.dropna(subset=["license"])

id                                NAME      host id  \
11114   7139598  Cozy 1 BR on Bedford Avenue, Wburg  73023181304   
72947  41289964  Cozy 1 BR on Bedford Avenue, Wburg  25804773951   

      host_identity_verified  host name neighbourhood group neighbourhood  \
11114               verified  Christina            Brooklyn  Williamsburg   
72947            unconfirmed  Christina            Brooklyn  Williamsburg   

            lat      long        country  ... service fee minimum nights  \
11114  40.71764 -73.95689  United States  ...       $140             1.0   
72947  40.71764 -73.95689  United States  ...       $140             1.0   

      number of reviews last review  reviews per month review rate number  \
11114               1.0    1/3/2016               0.02                1.0   
72947               1.0    1/3/2016               0.02                1.0   

      calculated host listings count  availability 365  \
11114                            1.0             191.0   
72947                            1.0               0.0   

                                             house_rules   license  
11114  Dear Guest, Thank you for appreciating that I ...  41662/AL  
72947                                                NaN  41662/AL  

[2 rows x 26 columns]

In [7]:
# Die Spalte license hat keine Aussagekraft, da es nur eine einzige Lizent gibt (bzw. weniger als 1 % der Daten haben eine Lizenz)
# Wir könnten diese Spalte also löschen

df = df.drop(columns=['license'], axis=1)

In [8]:
df[df["NAME"] == "Cozy 1 BR on Bedford Avenue, Wburg"]

id                                NAME      host id  \
11114   7139598  Cozy 1 BR on Bedford Avenue, Wburg  73023181304   
72947  41289964  Cozy 1 BR on Bedford Avenue, Wburg  25804773951   

      host_identity_verified  host name neighbourhood group neighbourhood  \
11114               verified  Christina            Brooklyn  Williamsburg   
72947            unconfirmed  Christina            Brooklyn  Williamsburg   

            lat      long        country  ...  price service fee  \
11114  40.71764 -73.95689  United States  ...  $702        $140    
72947  40.71764 -73.95689  United States  ...  $702        $140    

      minimum nights number of reviews  last review reviews per month  \
11114            1.0               1.0     1/3/2016              0.02   
72947            1.0               1.0     1/3/2016              0.02   

      review rate number  calculated host listings count  availability 365  \
11114                1.0                             1.0             191.0   
72947                1.0                             1.0               0.0   

                                             house_rules  
11114  Dear Guest, Thank you for appreciating that I ...  
72947                                                NaN  

[2 rows x 25 columns]

In [9]:
df[df["host name"] == "Christina"]

id                                               NAME  \
1263     1698891                     Entire loft, best Williamsburg   
1451     1802723                    Studio With Old-World Character   
2006     2109250                 Free Metrocard*, Safe & Affordable   
2007     2109802                     Private Suite, Free Metrocard*   
3327     2838838  Williamsburg, prime loction, Brooklyn private ...   
...          ...                                                ...   
101373  56989644  Great 1 BR- Q train, Museum Mile and Central P...   
101831  57242597                             Cozy Room in Manhattan   
102141  35552672                               Cozy Fully Furnished   
102348  20341777                                   Christina’s home   
102554   6070345                             Beautiful Private Room   

            host id host_identity_verified  host name neighbourhood group  \
1263     4038119569            unconfirmed  Christina            Brooklyn   
1451    44081645626            unconfirmed  Christina            Brooklyn   
2006    16777949923            unconfirmed  Christina              Queens   
2007    69900932774            unconfirmed  Christina              Queens   
3327     5677864921               verified  Christina            Brooklyn   
...             ...                    ...        ...                 ...   
101373  59039355401               verified  Christina           Manhattan   
101831   1436996918               verified  Christina           Manhattan   
102141  87897792397            unconfirmed  Christina            Brooklyn   
102348  89461531863               verified  Christina               Bronx   
102554  47602000155               verified  Christina            Brooklyn   

                    neighbourhood       lat      long        country  ...  \
1263                 Williamsburg  40.71872 -73.96042  United States  ...   
1451                Crown Heights  40.67189 -73.93499  United States  ...   
2006                     Flushing  40.75088 -73.81029  United States  ...   
2007                     Flushing  40.75250 -73.81126  United States  ...   
3327                 Williamsburg  40.71971 -73.95998  United States  ...   
...                           ...       ...       ...            ...  ...   
101373            Upper East Side  40.76868 -73.95744  United States  ...   
101831                     Harlem  40.82827 -73.94785  United States  ...   
102141  Prospect-Lefferts Gardens  40.65932 -73.95974  United States  ...   
102348                  Bronxdale  40.85460 -73.86785  United States  ...   
102554              East New York  40.67240 -73.88118  United States  ...   

          price service fee minimum nights number of reviews  last review  \
1263      $660        $132             4.0              33.0     4/7/2019   
1451       $92         $18             7.0               2.0    8/25/2015   
2006      $253         $51             1.0             116.0    6/14/2019   
2007      $933        $187             1.0             313.0     7/5/2019   
3327    $1,067        $213             3.0             176.0    6/21/2019   
...         ...         ...            ...               ...          ...   
101373    $448         $90             5.0              22.0     7/2/2019   
101831    $481         $96             1.0              17.0    10/8/2018   
102141    $303         $61            30.0              11.0    11/8/2021   
102348    $124         $25             1.0              24.0    6/27/2019   
102554    $533        $107            20.0              29.0    3/30/2019   

       reviews per month review rate number  calculated host listings count  \
1263                0.44                2.0                             2.0   
1451                0.03                5.0                             1.0   
2006                1.51                2.0                             2.0   
2007                4.09                2.0                           

In [10]:
print(df.drop_duplicates(subset='id').shape)
print(df.drop_duplicates(subset='NAME').shape)

(102058, 25)
(61282, 25)


In [11]:
print(df.drop_duplicates(subset='id').shape)
print(df.drop_duplicates(subset=['NAME', 'lat', 'long', 'host name']).shape)

(102058, 25)
(67747, 25)


In [12]:
print(df.drop_duplicates(subset=['NAME', 'lat', 'long', 'host name']))

              id                                               NAME  \
0        1001254                 Clean & quiet apt home by the park   
1        1002102                              Skylit Midtown Castle   
2        1002403                THE VILLAGE OF HARLEM....NEW YORK !   
3        1002755                                                NaN   
4        1003689   Entire Apt: Spacious Studio/Loft by central park   
...          ...                                                ...   
102045  57360790        Private Bedroom in Rego Park / Forest Hills   
102046  57361342             Sunny budget friendly room in Bushwick   
102047  57361894            Cozy Apartment on Bedford Ave, Brooklyn   
102048  57362446                Your 1 bedroom home away from home!   
102049  57362999  MASTER Cozy Bedroom Queen size 2 blocks Timesq...   

            host id host_identity_verified host name neighbourhood group  \
0       80014485718            unconfirmed  Madaline            Brookly

In [ ]:
# Hypothese: Die NAs sind großteils werte, die durch doppelte accounts zustande kommen
identify_na = df[df["host_identity_verified"].isna()][["NAME", "host name"]]
identify_not_na = df.dropna(subset=["host_identity_verified"])

identify_na.shape

(289, 2)

In [12]:
duplucated_host_identies = identify_not_na.merge(identify_na, on=["NAME", "host name"], how="inner")
duplucated_host_identies.head()

id                                             NAME      host id  \
0  2624546                    SAFE BUDGET IN  Brooklyn  NYC  59058469007   
1  2625098                 CHARM Rm HAMILTON HEIGHTS Harlem  16265048066   
2  2625650  Modern, Light-Filled Apt in Chelsea/Meatpacking  95051956475   
3  2626203                          The Top-Notch Top-Floor  20939120847   
4  2626755                BROOKLYN FAVORITE VACATION HOME 2  90154950485   

  host_identity_verified     host name neighbourhood group  neighbourhood  \
0            unconfirmed         Majar            Brooklyn  East New York   
1               verified         Shane           Manhattan         Harlem   
2            unconfirmed          Clay           Manhattan        Chelsea   
3               verified       Stephen            Brooklyn     Greenpoint   
4            unconfirmed  Mr. Williams            Brooklyn     Park Slope   

        lat      long        country  ... service fee minimum nights  \
0  40.66874 -73.88548  United States  ...       $148             3.0   
1  40.81976 -73.94029  United States  ...       $210             3.0   
2  40.74072 -74.00438  United States  ...       $100             7.0   
3  40.73370 -73.95427  United States  ...        $30             1.0   
4  40.68242 -73.97904  United States  ...       $183             3.0   

  number of reviews last review  reviews per month review rate number  \
0              34.0    4/6/2019               0.49                1.0   
1             101.0   6/10/2019               1.44                1.0   
2              11.0   6/22/2019               1.17                2.0   
3               0.0         NaN                NaN                1.0   
4              59.0   5/23/2019               0.84                1.0   

  calculated host listings count  availability 365  \
0                            8.0             184.0   
1                            1.0             330.0   
2                            1.0             371.0   
3                            1.0             137.0   
4                            1.0             384.0   

                                         house_rules license  
0                                                NaN     NaN  
1  NO PETS!  NO SMOKING ON PREMISES!  NO PARTY! A...     NaN  
2  Respect yourself, respect others, clean up aft...     NaN  
3                                                NaN     NaN  
4  Respect yourself respect others please clean u...     NaN  

[5 rows x 26 columns]

In [13]:
# Einfachster Umgang mit diesem Wert: Ersetze NA durch "unknown"
# Was machen wir: Werte, die nicht doppelt sind, durch "unknown" ersetzen

df[(df["host_identity_verified"].isna()) &
   (df[["NAME", "host name"]].isin(
       duplucated_host_identies[["NAME", "host name"]]))
   ]

: 

In [14]:
pd.set_option('display.max_rows', 100)


df[df['NAME'].isin(df[df['host_identity_verified'].isna()]
                   ['NAME'].tolist())].sort_values(['NAME', 'lat', 'long'])

# 1. nur host_identity_verified die NaN sind
# 2. gib nur den Namen zurück
# 3. Betrachte diese im gesamtdataframe
# 4. Sortiere nach dame
# Name sollte nun doppelt da sein, wenn der Host 2 Accounts haben könnte

id                                               NAME  \
287     1159846                     **Fantastic Williamsburg Apt**   
27399  16133804  *Beautiful 2BR*Flatiron*Best Location*Near Sub...   
467     1259260  *SoHo: Clean, Safe, Private, Peaceful Bedroom ...   
8430    5657225                   1 bedroom apt in pvt 2 fam house   
57517  32767975          1-Bdrm Newly Renovated Spacious Apartment   
...         ...                                                ...   
85894  48440592                                                NaN   
91996  51810727                                                NaN   
65867  37379680                                                NaN   
99879  56164508                                                NaN   
59499  33862634                                                NaN   

           host id host_identity_verified host name neighbourhood group  \
287    98480978167                    NaN     Baker                 NaN   
27399  81701629571                    NaN      Muni           Manhattan   
467    58284792631                    NaN    Stella           Manhattan   
8430    1864686359                    NaN     Debra            Brooklyn   
57517  64217290384               verified      Nick       Staten Island   
...            ...                    ...       ...                 ...   
85894  39843366779            unconfirmed    Andrea               Bronx   
91996  31116694016               verified     Julia               Bronx   
65867  64406215333               verified     Tanya               Bronx   
99879  43313998021            unconfirmed      Rony               Bronx   
59499  33512606255               verified    Andrea               Bronx   

        neighbourhood       lat      long        country  ...    price  \
287      Williamsburg  40.71031 -73.95830  United States  ...    $387    
27399         Midtown  40.74172 -73.98343  United States  ...     $51    
467              SoHo  40.72351 -73.99683  United States  ...    $817    
8430    Crown Heights  40.67255 -73.93328  United States  ...    $335    
57517     Great Kills  40.56071 -74.14591  United States  ...    $847    
...               ...       ...       ...            ...  ...      ...   
85894         Fordham  40.86032 -73.88493  United States  ...  $1,030    
91996  Pelham Gardens  40.86437 -73.84850  United States  ...    $148    
65867         Norwood  40.87553 -73.88601  United States  ...     $72    
99879     Kingsbridge  40.88297 -73.90727  United States  ...    $148    
59499       Wakefield  40.88332 -73.85018  United States  ...    $870    

      service fee minimum nights number of reviews  last review  \
287          $77             NaN               9.0    2/16/2019   
27399        $10             4.0              63.0    5/25/2019   
467         $163             2.0              54.0    6/16/2019   
8430         $67             1.0              24.0    6/22/2019   
57517       $169            14.0              77.0    1/10/2022   
...           ...            ...               ...          ...   
85894       $206             1.0               0.0          NaN   
91996        $30             2.0              40.0    4/16/2019   
65867        $14             3.0              24.0    3/25/2021   
99879        $30             3.0              54.0    6/24/2019   
59499       $174             3.0               7.0     1/1/2022   

      reviews per month review rate number  calculated host listings count  \
287                0.11                4.0                             1.0   
27399              3.17                5.0                             1.0   
467                0.56                1.0                             1.0   
8430               0.51                2.0                             1.0   
57517              1.38                2.0                             1.0   
...                 ...                ...                             ...   
85894               NaN        